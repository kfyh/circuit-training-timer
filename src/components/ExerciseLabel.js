import React from 'react'
import PropTypes from 'prop-types'

function ExerciseLabel (props) {
  const interval = props.interval
  const excerciseLabel = interval.label
  const totalReps = interval.reps

  return (
    <div>{excerciseLabel} ({props.currentRep}/{totalReps})</div>
  )
}

ExerciseLabel.propTypes = {
  interval: PropTypes.object,
  currentRep: PropTypes.number
}

export default ExerciseLabel
