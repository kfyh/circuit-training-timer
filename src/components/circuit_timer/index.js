export * from './StartStopButton.js'
export * from './ExerciseLabel.js'
export * from './TimeLabel.js'
