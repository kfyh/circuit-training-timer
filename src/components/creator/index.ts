export * from './AddExercisePage';
export * from './EditExercisePage';
export * from './ExerciseForm';
