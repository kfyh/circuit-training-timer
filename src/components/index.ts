export * from './CircuitTimer';
export * from './Selector';
export * from './timer/index';
export * from './Dashboard';
