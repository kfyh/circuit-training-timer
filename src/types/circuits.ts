export type Step = {
	label: string;
	time: number;
	reps: number;
	type: string;
	steps?: Array<Step>;
};

export type StepCircuit = Array<Step>;

export const NullStep = {
	label: 'null',
	time: 0,
	reps: 0,
	type: 'null',
	steps: undefined,
};

export type FlattenedStep = {
	id: string;
	label: string;
	time: number;
	repIndex: number;
	repTotal: number;
	type: string;
	parentId?: string;
	parentLabel?: string;
	parentType?: string;
	parentRepIndex?: number;
	parentRepTotal?: number;
};

export type FlattenedStepCircuit = Array<FlattenedStep>;

export const NullFlattenedStep = {
	id: '',
	label: 'null',
	time: 0,
	repIndex: 0,
	repTotal: 0,
	type: 'null',
	parentId: undefined,
	parentLabel: undefined,
	parentType: undefined,
	parentRepIndex: undefined,
	parentRepTotal: undefined,
};

export type Exercise = {
	id: string;
	name: string;
	description?: string;
};

export type ExerciseGroup = {
	name: string;
	exercises: Array<{
		exerciseId: string;
		duration: number;
		count: number;
		rest: number;
	}>;
	repetitions: number;
	rest: number;
};

export type Circuit = {
	id: string;
	name: string;
	exerciseGroups: Array<ExerciseGroup>;
	repetitions: number;
};

export type FlattenedCircuit = Array<FlattenedExercise>;

export type FlattenedExercise = {
	id: string;
	name: string;
	description?: string;
	duration: number;
	count?: number;
	groupName: string;
	groupRepIndex: number;
	groupRepetitions: number;
	circuitId: string;
	circuitName: string;
	circuitRepIndex: number;
	circuitRepetition: number;
	rest: number;
};

export const NullCircuit: Circuit = {
	id: 'nullId',
	name: 'null',
	exerciseGroups: [],
	repetitions: 1,
};

export const NullExercise: Exercise = {
	id: '0',
	name: 'null',
};

export const NullFlattenedExercise: FlattenedExercise = {
	id: '',
	name: 'null',
	duration: 0,
	groupName: 'null',
	groupRepIndex: 0,
	groupRepetitions: 0,
	circuitId: '',
	circuitName: 'null',
	circuitRepIndex: 0,
	circuitRepetition: 0,
	rest: 0,
};
