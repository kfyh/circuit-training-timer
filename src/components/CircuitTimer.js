import React from 'react'
import PropTypes from 'prop-types'
import { StartStopButton, TimeLabel, ExerciseLabel } from './circuit_timer'
import { StepIntervalController, SetIntervalController } from '../controllers'

class CircuitTimer extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      controller: null,
      currentIntervalIndex: 0,
      currentIntervalRep: 1,
      currentStepIndex: 0,
      currentStepRep: 1,
      isRunning: false,
      endTime: 0,
      timeLeft: 0
    }
  }

  componentDidMount () {
  }

  componentWillUnmount () {
    this.stopTimer()
  }

  render () {
    let currentInterval = this.props.intervals[this.state.currentIntervalIndex]
    let currentRep = this.state.currentIntervalRep
    if (currentInterval.type === 'set') {
      currentInterval = currentInterval.steps[this.state.currentStepIndex]
      currentRep = this.state.currentStepRep
    }

    return (
      <div>
        <ExerciseLabel interval={currentInterval} currentRep={currentRep} />
        <TimeLabel time={this.state.timeLeft}/>
        <StartStopButton isRunning={this.state.isRunning} onClick={this.handleClick} />
      </div>
    )
  }

  handleClick = () => {
    if (this.state.isRunning) {
      this.stopTimer()
    } else {
      this.startTimer()
    }
  }

  startTimer = () => {
    if (this.state.isRunning) {
      return
    }
    let controller = this.state.controller
    let timeLeft = this.state.timeLeft
    if (controller === null) {
      const interval = this.props.intervals[this.state.currentIntervalIndex]
      if (interval.type === 'set') {
        controller = new SetIntervalController(interval)
      } else {
        controller = new StepIntervalController(interval)
      }
      controller.start(Date.now())
      timeLeft = controller.timeLeft
    } else {
      controller.resume(Date.now())
      timeLeft = controller.timeLeft
    }

    this.setState({
      controller: controller,
      isRunning: true,
      endTime: Date.now() + timeLeft
    })

    this.timer = setInterval(this.update, 100)
  }

  stopTimer = () => {
    if (this.state.isRunning) {
      this.setState({
        isRunning: false
      })
      this.state.controller.pause()
      clearInterval(this.timer)
    }
  }

  update = () => {
    const values = this.state.controller.update(Date.now())
    if (values.isComplete) {
      const currentIntervalIndex = this.state.currentIntervalIndex + 1
      if (currentIntervalIndex >= this.props.intervals.length) {
        clearInterval(this.timer)
        this.setState({
          controller: null,
          currentIntervalIndex: 0,
          currentIntervalRep: 1,
          currentStepIndex: 0,
          currentStepRep: 1,
          isRunning: false,
          endTime: 0,
          timeLeft: 0
        })
        this.props.onComplete()
      } else {
        const nextInterval = this.props.intervals[currentIntervalIndex]
        let nextController
        if (nextInterval.type === 'set') {
          nextController = new SetIntervalController(nextInterval)
        } else {
          nextController = new StepIntervalController(nextInterval)
        }
        nextController.start(Date.now())
        this.setState({
          controller: nextController,
          timeLeft: nextController.timeLeft,
          currentIntervalIndex: currentIntervalIndex,
          currentIntervalRep: nextController.reps,
          endTime: nextController.endTime
        })
      }
    } else {
      this.setState({
        timeLeft: values.timeLeft,
        currentIntervalRep: values.reps,
        currentStepIndex: values.currentStepIndex ? values.currentStepIndex : 0
      })
    }
  }
}

CircuitTimer.propTypes = {
  intervals: PropTypes.arrayOf(PropTypes.object),
  onComplete: PropTypes.func
}

export default CircuitTimer
