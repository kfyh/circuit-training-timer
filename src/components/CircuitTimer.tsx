import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { History } from 'history';
import { StartStopButton, TimeLabel, ExerciseLabel } from './timer';
import { FlattenedCircuitController } from '../controllers';
import { FlattenedExercise, FlattenedCircuit } from '../types/circuits';
import { ISelectorReducerState } from '../reducers/selectorReducer';
import { flattenCircuit } from '../transforms/flattenTransform';

type CircuitTimerProps = {
	flattenedCircuit: FlattenedCircuit;
	history: History;
};

type CircuitTimerState = {
	currentIndex: number;
	hasStarted: boolean;
	isRunning: boolean;
	isResting: boolean;
	endTime: number;
	timeLeft: number;
};
export class CircuitTimer extends React.Component<CircuitTimerProps, CircuitTimerState> {
	private controller: FlattenedCircuitController;

	constructor(props: CircuitTimerProps) {
		super(props);

		this.controller = new FlattenedCircuitController(props.flattenedCircuit);
		this.state = {
			currentIndex: 0,
			hasStarted: false,
			isRunning: false,
			isResting: false,
			endTime: 0,
			timeLeft: 0,
		};
	}

	public componentWillUnmount(): void {
		this.stopTimer();
	}

	public render(): ReactElement {
		const exercise: FlattenedExercise = this.props.flattenedCircuit[this.state.currentIndex];

		return (
			<div>
				<ExerciseLabel exercise={exercise} isResting={this.state.isResting} />
				<TimeLabel time={this.state.timeLeft} isResting={this.state.isResting} />
				<StartStopButton isRunning={this.state.isRunning} onClick={this.handleClick} />
				<Link to="/">Home</Link>
			</div>
		);
	}

	private handleClick = (): void => {
		if (this.state.isRunning) {
			this.stopTimer();
		} else {
			this.startTimer();
		}
	};

	private startTimer = (): void => {
		if (this.state.isRunning) {
			return;
		}
		const controller = this.controller;
		let timeLeft = this.state.timeLeft;
		if (this.state.hasStarted) {
			timeLeft = controller.start(Date.now());
			this.setState(
				(state: CircuitTimerState): CircuitTimerState => {
					return { ...state, hasStarted: true };
				}
			);
		} else {
			timeLeft = controller.resume(Date.now());
		}

		this.setState({
			isRunning: true,
			endTime: Date.now() + timeLeft,
		});

		window.requestAnimationFrame(this.update);
	};

	private stopTimer = (): void => {
		if (this.state.isRunning) {
			this.setState({
				isRunning: false,
			});
			this.controller.pause();
		}
	};

	private update = (): void => {
		if (!this.state.isRunning) {
			return;
		}

		const values = this.controller.update(Date.now());
		if (values.isComplete) {
			this.endTimer();
		} else {
			this.setState({
				timeLeft: values.timeLeft,
				currentIndex: values.index,
				isResting: values.isResting,
			});

			if (this.state.isRunning) {
				window.requestAnimationFrame(this.update);
			}
		}
	};

	private endTimer = (): void => {
		this.setState({
			currentIndex: 0,
			hasStarted: false,
			isRunning: false,
			isResting: false,
			endTime: 0,
			timeLeft: 0,
		});
		this.props.history.push('/');
	};
}

const mapStateToProps = (state: ISelectorReducerState) => {
	const { id } = useParams<{ id: string }>();
	let circuit = state.circuits.find((circuit) => circuit.id === id);
	circuit = circuit ? circuit : state.currentCircuit;
	return {
		flattenedCircuit: flattenCircuit(circuit, state.exercises),
	};
};

export const ConnectedCircuitTimer = connect(mapStateToProps)(CircuitTimer);
