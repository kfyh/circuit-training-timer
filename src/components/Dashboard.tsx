import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { ISelectorAction, ISelectorReducerState } from '../reducers';
import { Circuit, Exercise } from '../types/circuits';
import { Dispatch } from 'redux';
import { selectCircuit } from '../actions/selectorActions';
import { CircuitListView, ExerciseListView } from './selector/';
import { History } from 'history';
import { Link } from 'react-router-dom';

type DashboardProps = {
	exercises: Array<Exercise>;
	circuits: Array<Circuit>;
	history: History;
	selectCircuit: (circuit: Circuit) => void;
};

export class Dashboard extends React.Component<DashboardProps, Record<string, never>> {
	constructor(props: DashboardProps) {
		super(props);
	}

	private onCircuitClicked = (circuit: Circuit) => {
		this.props.selectCircuit(circuit);
		this.props.history.push(`/timer/${circuit.id}`);
	};

	private onExerciseClicked = () => {
		
	};

	private onEditExercise = (exercise: Exercise): void => {
		this.props.history.push(`/editexercise/${exercise.id}`);
	}

	public render(): ReactElement {
		return (
			<div>
				<p>Dashboard</p>
				<a>New Circuit</a>
				<CircuitListView circuits={this.props.circuits} onSelect={this.onCircuitClicked} />
				<div id="exercises">
					<h2>Exercises</h2>
					<Link to="/addexercise">Add Exercise</Link>
					<ExerciseListView exercises={this.props.exercises} onSelect={this.onExerciseClicked} onEdit={this.onEditExercise} />
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state: ISelectorReducerState) => {
	return {
		circuits: state.circuits,
		exercises: state.exercises,
	};
};

const mapDispatchToProps = (dispatch: Dispatch<ISelectorAction>) => ({
	selectCircuit: (circuit: Circuit) => dispatch(selectCircuit(circuit)),
});

export const ConnectedDashboard = connect(mapStateToProps, mapDispatchToProps)(Dashboard);
