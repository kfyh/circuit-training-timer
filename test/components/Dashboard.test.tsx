import React from 'react';
import { shallow } from 'enzyme';
import { createMemoryHistory } from 'history';
import { Dashboard } from '../../src/components/Dashboard';
import { Circuit } from '../../src/types/circuits';

const circuits: Array<Circuit> = [
	{
		id: '0',
		name: 'Circuit One',
		exerciseGroups: [],
		repetitions: 1,
	},
	{
		id: '1',
		name: 'Circuit Two',
		exerciseGroups: [],
		repetitions: 1,
	},
];
test('Given circuits, dashboard displays correctly', () => {
	const selectMock = jest.fn();
	const history = createMemoryHistory();
	const render = shallow(<Dashboard circuits={circuits} history={history} selectCircuit={selectMock} />);
	expect(render).toMatchSnapshot();
});

test.skip('When circuit selected then select Circuit called', () => {
	expect(true);
});

test.skip('When circuit selected then go to timer', () => {
	expect(true);
});
