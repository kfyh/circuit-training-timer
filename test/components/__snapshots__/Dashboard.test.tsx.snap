// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given circuits, dashboard displays correctly 1`] = `
<div>
  <p>
    Dashboard
  </p>
  <a>
    New Circuit
  </a>
  <NavLink
    exact={true}
    to="/addexercise"
  >
    Add Exercise
  </NavLink>
  <CircuitListView
    circuits={
      Array [
        Object {
          "exerciseGroups": Array [],
          "id": "0",
          "name": "Circuit One",
          "repetitions": 1,
        },
        Object {
          "exerciseGroups": Array [],
          "id": "1",
          "name": "Circuit Two",
          "repetitions": 1,
        },
      ]
    }
    onSelect={[Function]}
  />
  <ExerciseListView
    onEdit={[Function]}
    onSelect={[Function]}
  />
</div>
`;
